File: impl-blocks/proxy-common/src/main/java/ca/stellardrift/permissionsex/proxycommon/ProxyContextDefinition.java
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.stellardrift.permissionsex.proxycommon;

import ca.stellardrift.permissionsex.context.ContextDefinition;
import ca.stellardrift.permissionsex.subject.CalculatedSubject;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.spongepowered.configurate.serialize.Scalars;

import java.util.Objects;
import java.util.function.Consumer;

public final class ProxyContextDefinition extends ContextDefinition<Boolean> {
    public static final ProxyContextDefinition INSTANCE = new ProxyContextDefinition();

    private ProxyContextDefinition() {
        super("proxy");
    }

    @Override
    public @NonNull String serialize(final Boolean canonicalValue) {
        return String.valueOf(canonicalValue);
    }

    @Override
    public @Nullable Boolean deserialize(final String userValue) {
        return Scalars.BOOLEAN.tryDeserialize(userValue);
    }

    @Override
    public boolean matches(final Boolean ownVal, final Boolean testVal) {
        return Objects.equals(ownVal, testVal);
    }

    @Override
    public void accumulateCurrentValues(final CalculatedSubject subject, final Consumer<Boolean> consumer) {
        consumer.accept(true);
    }
}

File: impl-blocks/proxy-common/src/main/kotlin/ca/stellardrift/permissionsex/proxycommon/ProxyContextDefinition.kt
/*
 * PermissionsEx
 * Copyright (C) zml and PermissionsEx contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.stellardrift.permissionsex.proxycommon

import ca.stellardrift.permissionsex.context.ContextDefinition
import ca.stellardrift.permissionsex.subject.CalculatedSubject
import ninja.leaping.configurate.Types

/**
 * This context exists to mark the fact that a subject is operating in a proxy environment.
 */
object ProxyContextDefinition : ContextDefinition<Boolean>("proxy") {
    override fun serialize(userValue: Boolean): String = userValue.toString()
    override fun deserialize(canonicalValue: String): Boolean = Types.asBoolean(canonicalValue) ?: false
    override fun matches(ownVal: Boolean, testVal: Boolean): Boolean = ownVal == testVal
    override fun accumulateCurrentValues(subject: CalculatedSubject, consumer: (value: Boolean) -> Unit) =
        consumer(true)
}
